File: python/examples/03_mix_enhance.py
#!/usr/bin/env python3
"""
03_mix_enhance.py - Enhance Your Mix with the Tonn API

This example demonstrates the Mix Enhance feature which automatically
improves your stereo mix without needing individual stems.

What Mix Enhance does:
- Fixes loudness issues for streaming platforms
- Corrects stereo width problems
- Improves tonal balance
- Optionally applies mastering
- Can perform stem separation and processing

The workflow:
1. First create a PREVIEW (quick, free to experiment)
2. Review the preview
3. Then create the FULL enhancement

Usage:
    python 03_mix_enhance.py [audio_url]

If no URL provided, uses a demo track from the test bucket.
"""

import sys
import os
import time

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from shared import TonnClient


def start_enhance_preview(client: TonnClient, audio_url: str, musical_style: str = "POP") -> str:
    """
    Start a Mix Enhance preview job.
    
    Returns the task ID for tracking the job.
    """
    payload = {
        "mixReviveData": {
            "audioFileLocation": audio_url,
            "musicalStyle": musical_style,
            "isMaster": False,
            "fixClippingIssues": True,
            "fixDRCIssues": True,
            "fixStereoWidthIssues": True,
            "fixTonalProfileIssues": True,
            "fixLoudnessIssues": True,
            "applyMastering": True,
            "loudnessPreference": "STREAMING_LOUDNESS",
            "stemProcessing": False
        }
    }
    
    print("🎚️  Starting Mix Enhance preview...")
    response = client.post("/mixenhancepreview", payload)
    
    if response and not response.get("error"):
        task_id = response.get("mixrevive_task_id")
        print(f"✓ Preview task created. Task ID: {task_id}")
        return task_id
    else:
        error_msg = response.get("message") if response else "No response"
        print(f"❌ Failed to start preview: {error_msg}")
        return None


def start_full_enhance(client: TonnClient, audio_url: str, musical_style: str = "POP") -> str:
    """
    Start a full Mix Enhance job (uses credits).
    
    Returns the task ID for tracking the job.
    """
    payload = {
        "mixReviveData": {
            "audioFileLocation": audio_url,
            "musicalStyle": musical_style,
            "isMaster": False,
            "fixClippingIssues": True,
            "fixDRCIssues": True,
            "fixStereoWidthIssues": True,
            "fixTonalProfileIssues": True,
            "fixLoudnessIssues": True,
            "applyMastering": True,
            "loudnessPreference": "STREAMING_LOUDNESS",
            "stemProcessing": False
        }
    }
    
    print("🎚️  Starting full Mix Enhance...")
    response = client.post("/mixenhance", payload)
    
    if response and not response.get("error"):
        task_id = response.get("mixrevive_task_id")
        print(f"✓ Full enhance task created. Task ID: {task_id}")
        return task_id
    else:
        error_msg = response.get("message") if response else "No response"
        print(f"❌ Failed to start enhancement: {error_msg}")
        return None


def poll_enhanced_track(client: TonnClient, task_id: str, max_attempts: int = 70) -> dict:
    """Poll for the enhanced track result."""
    payload = {
        "mixReviveData": {
            "mixReviveTaskId": task_id
        }
    }
    
    return client.poll_for_result(
        endpoint="/retrieveenhancedtrack",
        payload=payload,
        result_key="revivedTrackTaskResults",
        max_attempts=max_attempts,
        poll_interval=5,
        processing_status_codes=(202, 404, 503)
    )


def download_results(client: TonnClient, results: dict, prefix: str) -> None:
    """Download the enhanced track and any stems."""
    # Main enhanced track
    preview_url = results.get("download_url_preview_revived")
    full_url = results.get("download_url_revived")
    
    download_url = full_url or preview_url
    if download_url:
        filename = f"{prefix}_enhanced.wav"
        client.download_file(download_url, filename)
    
    # Stems (if any)
    stems = results.get("stems", {})
    for stem_name, stem_url in stems.items():
        filename = f"{prefix}_stem_{stem_name}.wav"
        client.download_file(stem_url, filename)


def main():
    # Demo audio URL (or use command line argument)
    demo_url = "https://storage.googleapis.com/test-bucket-api-roex/album/audio_track_1.mp3"
    
    if len(sys.argv) > 1:
        audio_url = sys.argv[1]
        print(f"Using provided URL: {audio_url}")
    else:
        audio_url = demo_url
        print(f"Using demo track: {audio_url}")
    
    client = TonnClient()
    
    print("\n" + "=" * 60)
    print("STEP 1: Create Preview Enhancement")
    print("=" * 60)
    
    # Start preview
    preview_task_id = start_enhance_preview(client, audio_url, "HIPHOP_GRIME")
    if not preview_task_id:
        print("Failed to create preview. Exiting.")
        sys.exit(1)
    
    # Poll for preview results
    print("\n⏳ Waiting for preview to complete...")
    preview_results = poll_enhanced_track(client, preview_task_id)
    
    if not preview_results:
        print("Preview timed out or failed.")
        sys.exit(1)
    
    print("\n📊 Preview Results:")
    print(f"  Status: {preview_results.get('status', 'Unknown')}")
    
    # Download preview
    download_results(client, preview_results, "preview")
    
    print("\n" + "=" * 60)
    print("STEP 2: Create Full Enhancement")
    print("=" * 60)
    
    # Start full enhancement
    full_task_id = start_full_enhance(client, audio_url, "HIPHOP_GRIME")
    if not full_task_id:
        print("Failed to start full enhancement. Exiting.")
        sys.exit(1)
    
    # Poll for full results
    print("\n⏳ Waiting for full enhancement to complete...")
    full_results = poll_enhanced_track(client, full_task_id, max_attempts=50)
    
    if not full_results:
        print("Full enhancement timed out or failed.")
        sys.exit(1)
    
    print("\n📊 Full Enhancement Results:")
    print(f"  Status: {full_results.get('status', 'Unknown')}")
    
    # Download full results
    download_results(client, full_results, "enhanced")
    
    print("\n" + "=" * 60)
    print("✅ Mix Enhance Complete!")
    print("=" * 60)
    print("\nYour enhanced files have been saved to the current directory.")


if __name__ == "__main__":
    main()

File: nodejs/src/examples/03_mix_enhance.js
#!/usr/bin/env node
/**
 * 03_mix_enhance.js - Enhance Your Mix with the Tonn API
 *
 * This example demonstrates the Mix Enhance feature which automatically
 * improves your stereo mix without needing individual stems.
 *
 * Usage:
 *   node 03_mix_enhance.js [audio_url]
 *
 * If no URL provided, uses a demo track from the test bucket.
 */

import { TonnClient } from '../shared/client.js';

/**
 * Start a Mix Enhance preview job.
 */
async function startEnhancePreview(client, audioUrl, musicalStyle = 'POP') {
  const payload = {
    mixReviveData: {
      audioFileLocation: audioUrl,
      musicalStyle,
      isMaster: false,
      fixClippingIssues: true,
      fixStereoWidthIssues: true,
      fixTonalProfileIssues: true,
      fixLoudnessIssues: true,
      applyMastering: true,
      stemProcessing: false
    }
  };

  console.log('🎚️  Starting Mix Enhance preview...');
  const response = await client.post('/mixenhancepreview', payload);

  if (response && !response.error) {
    const taskId = response.mixrevive_task_id;
    console.log(`✓ Preview task created. Task ID: ${taskId}`);
    return taskId;
  } else {
    const errorMsg = response?.message || 'No response';
    console.error(`❌ Failed to start preview: ${errorMsg}`);
    return null;
  }
}

/**
 * Start a full Mix Enhance job.
 */
async function startFullEnhance(client, audioUrl, musicalStyle = 'POP') {
  const payload = {
    mixReviveData: {
      audioFileLocation: audioUrl,
      musicalStyle,
      isMaster: false,
      fixClippingIssues: true,
      fixStereoWidthIssues: true,
      fixTonalProfileIssues: true,
      fixLoudnessIssues: true,
      applyMastering: true,
      stemProcessing: false
    }
  };

  console.log('🎚️  Starting full Mix Enhance...');
  const response = await client.post('/mixenhance', payload);

  if (response && !response.error) {
    const taskId = response.mixrevive_task_id;
    console.log(`✓ Full enhance task created. Task ID: ${taskId}`);
    return taskId;
  } else {
    const errorMsg = response?.message || 'No response';
    console.error(`❌ Failed to start enhancement: ${errorMsg}`);
    return null;
  }
}

/**
 * Poll for enhanced track result.
 */
async function pollEnhancedTrack(client, taskId, maxAttempts = 70) {
  const payload = {
    mixReviveData: {
      mixReviveTaskId: taskId
    }
  };

  return client.pollForResult({
    endpoint: '/retrieveenhancedtrack',
    payload,
    resultKey: 'revivedTrackTaskResults',
    maxAttempts,
    pollInterval: 5000
  });
}

/**
 * Download results.
 */
async function downloadResults(client, results, prefix) {
  const previewUrl = results.download_url_preview_revived;
  const fullUrl = results.download_url_revived;
  const downloadUrl = fullUrl || previewUrl;

  if (downloadUrl) {
    const filename = `${prefix}_enhanced.wav`;
    await client.downloadFile(downloadUrl, filename);
  }

  // Download stems if any
  const stems = results.stems || {};
  for (const [stemName, stemUrl] of Object.entries(stems)) {
    const filename = `${prefix}_stem_${stemName}.wav`;
    await client.downloadFile(stemUrl, filename);
  }
}

async function main() {
  const demoUrl = 'https://storage.googleapis.com/test-bucket-api-roex/album/audio_track_1.mp3';
  const audioUrl = process.argv[2] || demoUrl;

  if (process.argv[2]) {
    console.log(`Using provided URL: ${audioUrl}`);
  } else {
    console.log(`Using demo track: ${audioUrl}`);
  }

  const client = new TonnClient();

  console.log('\n' + '='.repeat(60));
  console.log('STEP 1: Create Preview Enhancement');
  console.log('='.repeat(60));

  // Start preview
  const previewTaskId = await startEnhancePreview(client, audioUrl, 'HIPHOP_GRIME');
  if (!previewTaskId) {
    console.error('Failed to create preview. Exiting.');
    process.exit(1);
  }

  // Poll for preview results
  console.log('\n⏳ Waiting for preview to complete...');
  const previewResults = await pollEnhancedTrack(client, previewTaskId);

  if (!previewResults) {
    console.error('Preview timed out or failed.');
    process.exit(1);
  }

  console.log('\n📊 Preview Results:');
  console.log(`  Status: ${previewResults.status || 'Unknown'}`);

  // Download preview
  await downloadResults(client, previewResults, 'preview');

  console.log('\n' + '='.repeat(60));
  console.log('STEP 2: Create Full Enhancement');
  console.log('='.repeat(60));

  // Start full enhancement
  const fullTaskId = await startFullEnhance(client, audioUrl, 'HIPHOP_GRIME');
  if (!fullTaskId) {
    console.error('Failed to start full enhancement. Exiting.');
    process.exit(1);
  }

  // Poll for full results
  console.log('\n⏳ Waiting for full enhancement to complete...');
  const fullResults = await pollEnhancedTrack(client, fullTaskId, 50);

  if (!fullResults) {
    console.error('Full enhancement timed out or failed.');
    process.exit(1);
  }

  console.log('\n📊 Full Enhancement Results:');
  console.log(`  Status: ${fullResults.status || 'Unknown'}`);

  // Download full results
  await downloadResults(client, fullResults, 'enhanced');

  console.log('\n' + '='.repeat(60));
  console.log('✅ Mix Enhance Complete!');
  console.log('='.repeat(60));
  console.log('\nYour enhanced files have been saved to the current directory.');
}

main().catch(console.error);

